File: backend-C#/DoctorsOffice/Services/DoctorService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using DoctorsOffice.DbModels;
using DoctorsOffice.Dto;
using DoctorsOffice.Repositories;

namespace DoctorsOffice.Services
{
    public class DoctorService 
    {
        private readonly DoctorRepository _doctorRepository;
        private readonly VisitService _visitService;

        public DoctorService(DoctorRepository doctorRepository, VisitService visitService)
        {
            _doctorRepository = doctorRepository;
            _visitService = visitService;
        }
        
        public List<DoctorWithAvailableHours> FindDoctorsByFacilityIdAvailableForGivenDay(int facilityId, DateTime date)
        {
            return _doctorRepository
                .FindAllByFacilityId(facilityId)
                .Select(doctor => GetDoctorWithAvailableDaysForGivenDate(doctor, date))
                .ToList();
        }


        private DoctorWithAvailableHours GetDoctorWithAvailableDaysForGivenDate(Doctor doctor, DateTime date)
        {
            var visitsForDoctor = _visitService.GetVisitsForDoctor(doctor.Pwz, date);

            if (visitsForDoctor.Count == 0)
            {
                return new DoctorWithAvailableHours()
                    {Doctor = doctor, AvailableHours = _visitService.CreateNoVisitHours()};
            }
            else
            {
                return new DoctorWithAvailableHours()
                {
                    Doctor = doctor,
                    AvailableHours = _visitService.AvailableHoursFromGivenVisits(visitsForDoctor)
                };
            }
        }
    }
}

File: backend-Java/src/main/java/pl/damkac/doctor/DoctorService.java
package pl.damkac.doctor;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import pl.damkac.visit.Visit;
import pl.damkac.visit.VisitService;

import java.time.LocalDate;
import java.util.List;
import java.util.NoSuchElementException;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
class DoctorService {

    private final DoctorRepository doctorRepository;
    private final VisitService visitService;

    public List<DoctorWithAvailableHours> findDoctorsByFacilityIdAvailableForGivenDay(Long facilityId, LocalDate date) {
        return doctorRepository
                .findAllByFacilityId(facilityId)
                .orElseThrow(() -> new NoSuchElementException("Could not found doctors for facility with id " + facilityId))
                .stream()
                .map(doctor -> getDoctorWithAvailableDaysForGivenDate(doctor, date))
                .collect(Collectors.toList());
    }

    public DoctorWithAvailableHours getDoctorWithAvailableDaysForGivenDate(Doctor doctor, LocalDate date){
        List<Visit> visitsForDoctor = visitService.getVisitsForDoctor(doctor.getPwz(), date);

        if(visitsForDoctor.isEmpty()){
            return DoctorWithAvailableHours
                    .builder()
                    .doctor(doctor)
                    .availableHours(visitService.createNoVisitHours())
                    .build();
        }
        else {
            return DoctorWithAvailableHours
                    .builder()
                    .doctor(doctor)
                    .availableHours(visitService.availableHoursFromGivenVisits(visitsForDoctor))
                    .build();
        }

    }
}
